crates/sqlbuiltins/src/functions/bson.rs

```diff
@@ -0,0 +1,177 @@
+use datasources::object_store::generic::GenericStoreAccess;
+use datasources::object_store::ObjStoreAccess;
+use protogen::metastore::types::catalog::RuntimePreference;
+
+use crate::functions::table_location_and_opts;
+
+#[derive(Debug, Clone, Copy, Default)]
+pub struct BsonScan {}
+
+#[async_trait]
+impl TableFunc for BsonScan {
+    fn runtime_preference(&self) -> RuntimePreference {
+        RuntimePreference::Unspecified
+    }
+
+    fn name(&self) -> &str {
+        "read_bson"
+    }
+
+    // TODO: in addition to a much needed refactor, most of this should be implemented as a
+    // TableProvider in the datasources bson package and just wrapped here.
+    async fn create_provider(
+        &self,
+        ctx: &dyn TableFuncContextProvider,
+        args: Vec<FuncParamValue>,
+        mut opts: HashMap<String, FuncParamValue>,
+    ) -> Result<Arc<dyn TableProvider>, ExtensionError> {
+        let (source_url, storage_options) = table_location_and_opts(ctx, args, &mut opts)?;
+
+        let store_access = GenericStoreAccess::new_from_location_and_opts(
+            source_url.to_string().as_str(),
+            storage_options,
+        )
+        .map_err(|e| ExtensionError::Arrow(e.into()))?;
+
+        let store = store_access
+            .create_store()
+            .map_err(|e| ExtensionError::Arrow(e.into()))?;
+        let path = source_url.path();
+
+        // assume that the file type is a glob and see if there are
+        // more files...
+        let mut list = store_access
+            .list_globbed(&store, path.as_ref())
+            .await
+            .map_err(|e| ExtensionError::Arrow(e.into()))?;
+
+        if list.is_empty() {
+            return Err(ExtensionError::String(format!(
+                "no matching objects for '{path}'"
+            )));
+        }
+
+        // for consistent results, particularly for the sample, always
+        // sort by location
+        list.sort_by(|a, b| a.location.cmp(&b.location));
+
+        // parse arguments to see how many documents to consider
+        let sample_size = match opts.get("schema_sample_size") {
+            // TODO: set a maximum (1024?) or have an adaptive mode
+            // (at least n but stop after n the same) or skip documents
+            Some(v) => v.to_owned().param_into()?,
+            None => 100,
+        };
+
+        // by default we'll set up streams that could, in theory (after the sample) read from the
+        // files in parallel, leading to a table that will be streamed
+        let _serial_scan = opts.get("serial_scan").is_some();
+
+        // build a vector of streams, one for each file, that handle BSON's framing.
+        let mut readers = VecDeque::<
+            futures::stream::Map<
+                FramedRead<BufReader, LengthDelimitedCodec>,
+                fn(Result<BytesMut, std::io::Error>) -> Result<bson::Document, ExtensionError>,
+            >,
+        >::with_capacity(list.len());
```

Comment:
```suggestion
        let mut readers = VecDeque::with_capacity(list.len());
```

Could just make it inferred to reduce verbosity.

Reply:
oh interesting; were that I knew that before. 

It did actually help to put the type out like this before while I was writing.